pkg/github/enterpriseuserwriter.go

```diff
@@ -0,0 +1,128 @@
+type EnterpriseUserWriter struct {
+	scimClient             *SCIMClient
+	max_users_to_provision int64
+}
+
+// NewEnterpriseUserWriter creates a new EnterpriseUserWriter with default 1000
+// maximum number of users to provision if not override by given opts.
+func NewEnterpriseUserWriter(httpClient *http.Client, enterpriseBaseURL string, opts ...EnterpriseRWOpt) (*EnterpriseUserWriter, error) {
+	scimClient, err := NewSCIMClient(httpClient, enterpriseBaseURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create scim client: %w", err)
+	}
+	w := &EnterpriseUserWriter{
+		max_users_to_provision: defaultMaxUsersToProvision,
+		scimClient:             scimClient,
+	}
+	for _, opt := range opts {
+		opt(w)
+	}
+	return w, nil
+}
+
+// SetMembers creates and suspends enterprise users given the desired members.
+func (w *EnterpriseUserWriter) SetMembers(ctx context.Context, _ string, members []groupsync.Member) error {
+	logger := logging.FromContext(ctx)
+
+	currentUsersMap, err := w.scimClient.ListUsers(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to list users: %w", err)
+	}
+	desiredUsersMap := make(map[string]*github.SCIMUserAttributes)
+	for _, m := range members {
+		if !m.IsUser() {
+			logger.DebugContext(ctx, "skipping non-user member", "member", m.ID())
+			continue
+		}
+		u, _ := m.User()
+		scimUser, ok := u.Attributes.(*github.SCIMUserAttributes)
+		if !ok {
+			logger.DebugContext(ctx, "skipping non-SCIM user member", "member", m.ID())
+			continue
+		}
+		desiredUsersMap[scimUser.UserName] = scimUser
+	}
+
+	var merr error
+	// 1. Deactivate users to free up license seats first.
+	for username, scimUser := range currentUsersMap {
+		if _, ok := desiredUsersMap[username]; !ok {
+			if scimUser.Active != nil && !*scimUser.Active {
+				continue
+			}
+
+			logger.InfoContext(ctx, "deactivating user", "user", username)
+			if _, _, err := w.scimClient.DeactivateUser(ctx, *scimUser.ID); err != nil {
+				merr = errors.Join(merr, fmt.Errorf("failed to deactivate %q: %w", username, err))
+			}
+		}
+	}
+
+	// 2. Create and reactivate users.
+	var count int64
+	for username, desiredUser := range desiredUsersMap {
+		count++
+		if count > w.max_users_to_provision {
+			merr = errors.Join(merr, fmt.Errorf("exceeded max users to provision: %d", w.max_users_to_provision))
+			break
+		}
+		// Check if the user need to be reactived.
+		if _, ok := currentUsersMap[username]; ok {
+			if currentUsersMap[username].Active != nil && *currentUsersMap[username].Active {
+				continue
+			}
+			if _, _, err := w.scimClient.ReactivateUser(ctx, *currentUsersMap[username].ID); err != nil {
+				merr = errors.Join(merr, fmt.Errorf("failed to reactivate %q: %w", username, err))
+			}
+			continue
+		}
+		// Create User
+		if _, _, err := w.scimClient.CreateUser(ctx, desiredUser); err != nil {
+			merr = errors.Join(merr, fmt.Errorf("failed to create %q: %w", username, err))
```

Comment:
How about moving the body of this for loop into a separate function? That would save a little complexity around errors.Join and `continue`